common/src/main/java/io/pravega/common/io/ByteBufferOutputStream.java

```diff
@@ -0,0 +1,183 @@
+    }
+
+    @Override
+    public void write(byte[] array, int arrayOffset, int length, int streamPosition) {
+        if (arrayOffset < 0 || length < 0 || (length > 0 && arrayOffset + length > array.length)) {
+            throw new ArrayIndexOutOfBoundsException("bufferOffset and length must refer to a range within buffer.");
+        }
+
+        final int extra = streamPosition + length - this.buf.position();
+        if (extra > 0) {
+            ensureExtraCapacity(extra);
+        }
+
+        final int originalPos = this.buf.position();
+        this.buf.position(streamPosition);
+        this.buf.put(array, arrayOffset, length);
+        this.buf.position(Math.max(originalPos, streamPosition + length));
+    }
+
+
+    @Override
+    public int size() {
+        return this.buf.position();
+    }
+
+    @Override
+    public void flush() {
+        // This method intentionally left blank.
+    }
+
+    @Override
+    public void close() {
+        // This method intentionally left blank.
+    }
+
+    //endregion
+
+    //region RandomAccessOutputStream and DirectDataOutput Implementation
+
+    @Override
+    public void writeBuffer(BufferView buffer) {
+        ensureExtraCapacity(buffer.getLength());
+        int bytesCopied = buffer.copyTo(this.buf);
+        assert bytesCopied == buffer.getLength();
+    }
+
+    @Override
+    public void writeShort(int shortValue) {
+        ensureExtraCapacity(Short.BYTES);
+        this.buf.putShort((short) shortValue);
+    }
+
+    @Override
+    public void writeInt(int intValue) {
+        ensureExtraCapacity(Integer.BYTES);
+        this.buf.putInt(intValue);
+    }
+
+    @Override
+    public void writeInt(int intValue, int streamPosition) {
+        this.buf.putInt(streamPosition, intValue);
+    }
+
+    @Override
+    public void writeLong(long longValue) {
+        ensureExtraCapacity(Long.BYTES);
+        this.buf.putLong(longValue);
+    }
+
+    @Override
+    public ByteArraySegment getData() {
+        return new ByteArraySegment(this.buf.array(), this.buf.arrayOffset(), size());
+    }
+
+    //endregion
+
+    //region Helpers
+
+    private void ensureExtraCapacity(int count) {
+        // This method is borrowed from ByteArrayOutputStream and adapted for use with ByteBuffers.
+        final int minCapacity = this.buf.position() + count;
+        if (minCapacity <= this.buf.limit()) {
+            return;
+        }
+
+        final int oldCapacity = this.buf.limit();
+        int newCapacity = oldCapacity << 1;
+        if (newCapacity - minCapacity < 0) {
+            newCapacity = minCapacity;
+        }
+
+        if (newCapacity - MAX_ARRAY_SIZE > 0) {
+            if (minCapacity < 0) {
+                throw new OutOfMemoryError();
+            } else {
+                newCapacity = minCapacity > MAX_ARRAY_SIZE ? Integer.MAX_VALUE : MAX_ARRAY_SIZE;
+            }
+        }
+
+        final int oldPosition = this.buf.position();
+        this.buf.position(0);
```

Comment:
Have you thought about releasing the old ByteBuffer with the cleaner ?

Reply:
See above